scipy/stats/stats.py

```diff
@@ -2465,15 +2465,18 @@ def gstd(a, axis=0, ddof=1):
             raise ValueError(
                 'Infinite value encountered. The geometric standard deviation '
                 'is defined for strictly positive values only.')
-        elif np.less_equal(a, 0).any():
+        a_nan = np.isnan(a)
+        a_nan_any = a_nan.any()
+        if ((a_nan_any and np.less_equal(a[~a_nan], 0).any()) or
+              (not a_nan_any and np.less_equal(a, 0).any())):
```

Comment:
Do the two cases (`a` contains `NAN`s, `a` does not contain `NAN`s) need to be treated separately in the `if` clause?  I.e. Is this following snippet equivalent? (It also may remove the need for the `a_nan_any` variable)
```
a_nan = np.isnan(a)
if np.less_equal(a[~a_nan], 0).any():
    raise ValueError(
        'Non positive value encountered. The geometric standard '
	'deviation is defined for strictly positive values only.')
```

Reply:
Yes, for efficiency sake.

```
In [1]: import numpy as np                                                                                                                                                                                                                   

In [3]: a = np.take([1.0, 2.0, 3.0], np.random.choice(3, 10**6))                                                                                                                                                                             

In [4]: %timeit np.less(a, 0).any()                                                                                                                                                                                                          
418 µs ± 2.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]: a_nan = np.isnan(a)                                                                                                                                                                                                                  

In [7]: %timeit np.less(a[~a_nan], 0).any()                                                                                                                                                                                                  
2.86 ms ± 89.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

If there are no nans present, we do not want to run mask selection.

Reply:
Thanks.  The difference in time is small for `a` with  a thousand elements but is quite substantial by the time `a` has a million elements. Can you add a comment in the code indicating that's why the calculation s being structured in this manner?

Reply:
It is actually even better to avoid computing `np.less(a[~a_nan], 0)` at all, and do 
`np.less(np.nanmin(a), 0)` instead:

```
In [6]: a = np.take([1.0, np.nan, 3.0], np.random.choice(3, 10**6))                                                                                                                                                                          

In [7]: %timeit np.less(np.nanmin(a), 0)                                                                                                                                                                                                     
4.75 ms ± 9.82 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: a_nan = np.isnan(a)                                                                                                                                                                                                                  

In [9]: %timeit np.less(a[~a_nan], 0).any()                                                                                                                                                                                                  
8.23 ms ± 7.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: (np.less(a[~a_nan], 0).any(), np.less(np.nanmin(a), 0))                                                                                                                                                                             
Out[10]: (False, False)
```

It `np.nanmin` is a little slower than bare `np.less` for array of normal reals:

```
In [11]: a = np.take([1.0, 2.0, 3.0], np.random.choice(3, 10**6))                                                                                                                                                                            

In [12]: %timeit np.less(np.nanmin(a), 0)                                                                                                                                                                                                    
777 µs ± 922 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [13]: %timeit np.less(a, 0).any()                                                                                                                                                                                                         
429 µs ± 363 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
I like that. 
`np.less_equal(a, 0).any()` warns if it encounters `np.nan`, and expecting it `less_equal` to correctly decide that `np.nan < 0` is `False` may be asking a bit much on all platforms.
`np.nanmin(a)` avoids the warning.  It's definitely slower than `np.min(a)` so separating the two paths seems reasonable, even in a catch Exception block, as an array containing `np.nan` may be quite a common occurrence.
[This does raise the question as to what arrays might generate a `RuntimeWarning` but don't contain negative or `nan` values and hence get called a second time at the end of the catch block, with the same arguments as the first time.]